### PR TITLE
Tweak synchronous memcpy implementation

### PR DESCRIPTION
**Pull request by AlexVlx**
The existing one can have issues on certain systems, therefore this limits use of direct memcpy via largeBAR to sizes where it is unequivocally better.

Also addresses SWDEV-220030.

### CODE REVIEW

**Comment by jeffdaily** [comment]
FYI jenkins build 15 failed and 17 failed, but build 16 succeeded.  I believe this PR is actually passing CI now if we could try kicking off another try.

**Comment by gargrahul** [comment]
> 
> 
> FYI jenkins build 15 failed and 17 failed, but build 16 succeeded. I believe this PR is actually passing CI now if we could try kicking off another try.

Done. #18 is in progress now.

**Comment by jeffdaily** [comment]
@gargrahul did something change recently?  I'm seeing this error in the CI logs:
Error: docker: Error response from daemon: Unknown runtime specified nvidia.

File: src/hip_memory.cpp [hunk 1]
```diff
@@ -204,6 +204,12 @@ namespace {
 inline
 void do_copy(void* __restrict dst, const void* __restrict src, size_t n,
              hsa_agent_t da, hsa_agent_t sa) {
+    if (da.handle == 0 || sa.handle == 0) { // Handle degenerate case.
+        throwing_result_check(hsa_memory_copy(dst, src, n),
+                              __FILE__, __func__, __LINE__);        
+        return;
+    }
```
Comment (SarbojitAMD) [inline src/hip_memory.cpp]: Not sure if we still need this change as we have observed hangs in hsa_memory_copy().

Comment (SarbojitAMD) [reply src/hip_memory.cpp]: Ignore the comment missed your latest commit :)

**Comment by jedwards-AMD** [comment]
Lets merge this in.

**Pull request merged**